### PR TITLE
fix(watchtower): deduplicate nostr discovery events with shared channel worker

### PR DESCRIPTION
**Pull request by ManthanNimodiya**
## Summary                                                                                                                                             
In market discovery (`src/watch_tower/nostr.rs`), each configured Nostr relay previously opened an independent blockchain connection and independently raced to query raw transactions from the node/Electrum for incoming fidelity announcements. When multiple relays broadcasted the same announcement, duplicate RPC queries hit the node simultaneously over multiple sockets.

This PR introduces a lightweight channel-based producer-consumer model:
- Relay reader threads decode frames and stream incoming messages into a shared channel (`crossbeam-channel`).
- A single processor worker consumes from the channel, deduplicates transactions via `SeenTxids` prior to making RPC calls, and validates using a single shared blockchain connection.
- Duplicate events across relays skip the blockchain query while correctly advancing the relay's cursor.

## Testing
- Added unit tests in `src/watch_tower/nostr.rs`:
- `duplicate_event_across_relays_skips_blockchain_and_advances_cursor`
- `eose_message_returns_true`
- Ran full test suite:

```bash
cargo test --lib
cargo fmt --check
cargo clippy --all-features --lib --bins --tests -- -D warnings
```

## Checklist

[✓] Tests were added or updated where needed
[✓] Formatting and lints pass
[ ] Documentation was updated where needed (or N/A)
[✓] Security or protocol impact is explained above, if applicable

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved Nostr discovery reliability across multiple relay connections.
  * Improved duplicate-event handling and initial synchronization behavior.
  * Improved recovery when individual messages or relay connections encounter errors.
  * Improved startup and partial-failure handling during relay discovery.
  * Improved validation and registry updates for discovered events.

* **Tests**
  * Added coverage for duplicate events, end-of-stream processing, worker disconnects, and relay discovery edge cases.

### CODE REVIEW

File: src/watch_tower/nostr.rs [hunk 1]
```diff
@@ -128,17 +152,55 @@ fn join_relay_sessions(sessions: Vec<std::thread::JoinHandle<()>>) {
...
+    let mut seen_txid = SeenTxids::new();
+
+    while !shutdown.load(Ordering::SeqCst) {
+        let (relay_url, msg) = match rx.recv_timeout(std::time::Duration::from_millis(100)) {
+            Ok(item) => item,
+            Err(crossbeam_channel::RecvTimeoutError::Timeout) => continue,
+            Err(crossbeam_channel::RecvTimeoutError::Disconnected) => break,
+        };
+
+        let is_eose = match handle_relay_message(
```
Comment (mojoX911) [inline src/watch_tower/nostr.rs]: **Should fix (availability): this worker is now a single point of failure for all relays.** If `handle_relay_message` panics on one poisoned event — relay input is untrusted — the receiver drops, every relay session's `send_timeout` returns `Disconnected` (L320), the sessions return `Ok` and break out of their reconnect loops (L217-220), and discovery stops for *every* relay until restart. Before this PR, one bad event only killed that relay's session while the others kept discovering. Wrap this dispatch in `std::panic::catch_unwind` (or respawn the worker on exit) so one bad event can't blind the whole offerbook.

File: src/watch_tower/nostr.rs [hunk 2]
```diff
@@ -483,4 +529,73 @@ mod tests {
         // A well-formed frame still decodes, so the skip is not swallowing everything.
         assert!(decode_relay_frame(Message::Text(r#"["EOSE","sub1"]"#.into()), relay).is_some());
     }
+
+    #[test]
+    fn duplicate_event_across_relays_skips_blockchain_and_advances_cursor() {
```
Comment (mojoX911) [inline src/watch_tower/nostr.rs:611]: **Should fix (tests): the new machinery itself is untested.** Both added tests drive `handle_relay_message` directly, which this PR barely changes — nothing fails if the worker is never spawned, never exits on channel disconnect, or never sets `initial_sync_complete` when it processes an EOSE. Add one unit test that spawns `run_discovery_event_processor` with a real channel and the dummy `CoreRpc` backend already used below: push an EOSE, assert the flag flips, drop the sender, and assert the thread exits.

File: src/watch_tower/nostr.rs [hunk 3]
```diff
@@ -267,18 +310,17 @@ fn read_event_loop(
...
-            blockchain.clone(),
-            relay_url,
-            kind,
-            seen_txid,
-        )?;
-
-        if is_eose && !initial_sync_complete.load(Ordering::SeqCst) {
-            initial_sync_complete.store(true, Ordering::SeqCst);
-            log::info!("Initial Nostr discovery sync complete (triggered by {relay_url})");
+        let mut to_send = (relay_url.to_string(), relay_msg);
```
Comment (mojoX911) [inline src/watch_tower/nostr.rs]: **Nit:** `relay_url.to_string()` allocates once per message per relay. Building an `Arc<str>` (or a boxed `String`) once per session and cloning it into the tuple sends the same data for free.

**Review by mojoX911 — CHANGES_REQUESTED** [review]
Security review of the dedup/channel refactor, trust-boundary focus.

**No blockers.** Fund safety is clean: the txid is still claimed before any RPC and released on transient fetch failure, the cursor skew guard is unchanged, and a relay can still force only one RPC per unique txid. The worker's shared connection is the discovery-dedicated, shutdown-armed one built in `watcher.rs`, and it has a single consumer, so the Electrum single-client wedge risk does not apply. The bounded queue (1024) is a strict improvement over the old duplicate-RPC race.

Two should-fix items and one nit inline. Requesting changes for the worker single-point-of-failure and the missing worker test — both are small.

**Comment by mojoX911** [comment]
Also needs a rebase a lint fix.

**Comment by ManthanNimodiya** [comment]
@mojoX911 Thanks for the review, addressed all points and rebased:

1. Worker Panic Guard: 
Wrapped message processing in catch_unwind(AssertUnwindSafe(...)) inside run_discovery_event_processor so a malformed event cannot crash the processor or disconnect discovery for other relays.
2. Worker Unit Test: 
Added worker_handles_eose_and_exits_on_disconnect to verify EOSE handling (initial_sync_complete) and clean worker termination on channel disconnect.
3. Allocation Optimization: 
Switched channel to (Arc<str>, RelayMessage<'static>) to eliminate per-message string allocations.
4. Rebase & Lints: 
Rebased on master, formatted with cargo fmt, verified clippy passes with -D warnings, and signed the commit.

Ready for review.

**Review by mojoX911 — APPROVED** [review]
Ack
